src/Streamly/Memory/Array/Types.hs

```diff
@@ -893,6 +900,40 @@ read = D.fromStreamD . toStreamD
 -- Semigroup and Monoid
 -------------------------------------------------------------------------------
 
+-- Splice two immutable arrays creating a new array.
+{-# INLINE spliceTwo #-}
+spliceTwo :: (MonadIO m, Storable a) => Array a -> Array a -> m (Array a)
+spliceTwo arr1 arr2 = do
+    let src1 = unsafeForeignPtrToPtr (aStart arr1)
+        src2 = unsafeForeignPtrToPtr (aStart arr2)
+        len1 = aEnd arr1 `minusPtr` src1
+        len2 = aEnd arr2 `minusPtr` src2
+
+    arr <- liftIO $ newArray (len1 + len2)
+    let dst = unsafeForeignPtrToPtr (aStart arr)
+
+    -- XXX Should we use copyMutableByteArray# instead? Is there an overhead to
+    -- ccall?
+    liftIO $ do
+        memcpy (castPtr dst) (castPtr src1) len1
+        touchForeignPtr (aStart arr1)
+        memcpy (castPtr (dst `plusPtr` len1)) (castPtr src2) len2
+        touchForeignPtr (aStart arr2)
+    return arr { aEnd = dst `plusPtr` (len1 + len2) }
+
+instance Storable a => Semigroup (Array a) where
+    arr1 <> arr2 = unsafeInlineIO $ spliceTwo arr1 arr2
```

Comment:
From my understanding, `unsafeInlineIO` should not be used if we are allocating memory. There is a chance that multiple calls to malloc are shared, resulting in different arrays sharing the same memory address. I don't know if this is not true in newer releases.

Reply:
Yes, theoretically GHC can potentially optimize/float multiple mallocs into one thinking they are perfectly pure computations, though in this case it should not be easy for the compiler to do it. However, to be sure we should definitely change it to `unsafeDupablePerformIO`.